File: mobile-sdk/react-native/uxcam-react-native/ios/UXCamModule.swift
//
//  UXCamModule.swift
//  UXCam React Native Module for iOS
//
//  Created for UXCam Analytics Platform
//

import Foundation
import UIKit

@objc(UXCamModule)
class UXCamModule: NSObject {
  
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  // Import React Native bridge
  // Note: In actual React Native project, this would be:
  // import React
  // and class would conform to RCTBridgeModule
  
  @objc
  func initialize(_ config: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let sdkKey = config["sdkKey"] as? String else {
      reject("INVALID_ARGUMENT", "sdkKey is required", nil)
      return
    }
    
    let apiUrl = config["apiUrl"] as? String ?? "https://architecture-hormone-specify-bureau.trycloudflare.com"
    let enableEventTracking = config["enableEventTracking"] as? Bool ?? true
    let enableAutomaticTracking = config["enableAutomaticTracking"] as? Bool ?? true
    
    let uxcamConfig = UXCamConfig(
      sdkKey: sdkKey,
      apiUrl: apiUrl,
      enableEventTracking: enableEventTracking,
      enableAutomaticTracking: enableAutomaticTracking
    )
    
    UXCamSDK.shared.initialize(config: uxcamConfig)
    resolve(true)
  }
  
  @objc
  func getSessionId(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    let sessionId = UXCamSDK.shared.getSessionId()
    resolve(sessionId)
  }
  
  @objc
  func trackScreenView(_ params: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let screenName = params["screenName"] as? String else {
      reject("INVALID_ARGUMENT", "screenName is required", nil)
      return
    }
    
    let properties = params["properties"] as? [String: Any] ?? [:]
    UXCamSDK.shared.trackPageView(page: screenName, properties: properties)
    resolve(true)
  }
  
  @objc
  func trackEvent(_ params: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let eventName = params["eventName"] as? String else {
      reject("INVALID_ARGUMENT", "eventName is required", nil)
      return
    }
    
    let properties = params["properties"] as? [String: Any] ?? [:]
    UXCamSDK.shared.trackEvent(eventName, properties: properties)
    resolve(true)
  }
  
  @objc
  func setUserProperties(_ params: NSDictionary, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    let userId = params["user_id"] as? String
    let properties = params["properties"] as? [String: Any] ?? [:]
    
    UXCamSDK.shared.setUserProperties(userId: userId, properties: properties)
    resolve(true)
  }
  
  @objc
  func startRecording(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    UXCamSDK.shared.startVideoRecording()
    resolve(true)
  }
  
  @objc
  func stopRecording(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    UXCamSDK.shared.stopVideoRecording { success in
      resolve(success)
    }
  }
  
  @objc
  func endSession(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    UXCamSDK.shared.endSession()
    resolve(true)
  }
  
  @objc
  func restartSession(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    UXCamSDK.shared.restartSession()
    resolve(true)
  }
}

// Note: In actual React Native project setup:
// 1. Add React Native framework import: import React
// 2. RCTPromiseResolveBlock and RCTPromiseRejectBlock are provided by React Native
// 3. The UXCamModule.m file provides the Objective-C bridge to React Native
// 4. This Swift file connects to UXCamSDK.swift (native iOS SDK)

File: mobile-sdk/react-native/uxcam-react-native/android/src/main/java/com/uxcam/reactnative/UXCamModule.java
package com.uxcam.reactnative;

import android.content.Context;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.Arguments;
import com.uxcam.sdk.UXCamConfig;
import com.uxcam.sdk.UXCamSDK;

public class UXCamModule extends ReactContextBaseJavaModule {
  private static final String MODULE_NAME = "UXCamModule";
  private final ReactApplicationContext reactContext;

  public UXCamModule(ReactApplicationContext reactContext) {
    super(reactContext);
    this.reactContext = reactContext;
  }

  @Override
  public String getName() {
    return MODULE_NAME;
  }

  @ReactMethod
  public void initialize(ReadableMap config, Promise promise) {
    try {
      String sdkKey = config.getString("sdkKey");
      String apiUrl = config.hasKey("apiUrl") ? config.getString("apiUrl") : "https://architecture-hormone-specify-bureau.trycloudflare.com";
      boolean enableVideoRecording = config.hasKey("enableVideoRecording") ? config.getBoolean("enableVideoRecording") : true;
      boolean enableEventTracking = config.hasKey("enableEventTracking") ? config.getBoolean("enableEventTracking") : true;
      boolean enableAutomaticTracking = config.hasKey("enableAutomaticTracking") ? config.getBoolean("enableAutomaticTracking") : true;

      UXCamConfig uxcamConfig = new UXCamConfig(
        sdkKey,
        apiUrl,
        enableVideoRecording,
        enableEventTracking,
        enableAutomaticTracking
      );

      Context context = reactContext.getApplicationContext();
      UXCamSDK.getInstance().initialize(context, uxcamConfig);
      
      promise.resolve(true);
    } catch (Exception e) {
      promise.reject("INIT_ERROR", e.getMessage(), e);
    }
  }

  @ReactMethod
  public void getSessionId(Promise promise) {
    try {
      String sessionId = UXCamSDK.getInstance().getSessionId();
      promise.resolve(sessionId);
    } catch (Exception e) {
      promise.reject("GET_SESSION_ERROR", e.getMessage(), e);
    }
  }

  @ReactMethod
  public void trackScreenView(ReadableMap params, Promise promise) {
    try {
      String screenName = params.getString("screenName");
      ReadableMap propertiesMap = params.hasKey("properties") ? params.getMap("properties") : null;
      
      java.util.Map<String, Object> properties = new java.util.HashMap<>();
      if (propertiesMap != null) {
        // Convert ReadableMap to Map
        // Note: This is simplified - you'd need to handle all types
        properties.put("screen", screenName);
      }
      
      UXCamSDK.getInstance().trackPageView(screenName, properties);
      promise.resolve(true);
    } catch (Exception e) {
      promise.reject("TRACK_ERROR", e.getMessage(), e);
    }
  }

  @ReactMethod
  public void trackEvent(ReadableMap params, Promise promise) {
    try {
      String eventName = params.getString("eventName");
      ReadableMap propertiesMap = params.hasKey("properties") ? params.getMap("properties") : null;
      
      java.util.Map<String, Object> properties = new java.util.HashMap<>();
      if (propertiesMap != null) {
        // Convert ReadableMap to Map
        // Note: This is simplified - you'd need to handle all types
      }
      
      UXCamSDK.getInstance().trackEvent(eventName, properties);
      promise.resolve(true);
    } catch (Exception e) {
      promise.reject("TRACK_ERROR", e.getMessage(), e);
    }
  }

  @ReactMethod
  public void setUserProperties(ReadableMap params, Promise promise) {
    try {
      String userId = params.hasKey("user_id") ? params.getString("user_id") : null;
      ReadableMap propertiesMap = params.hasKey("properties") ? params.getMap("properties") : null;
      
      java.util.Map<String, Object> properties = new java.util.HashMap<>();
      if (propertiesMap != null) {
        // Convert ReadableMap to Map
      }
      
      UXCamSDK.getInstance().setUserProperties(userId, properties);
      promise.resolve(true);
    } catch (Exception e) {
      promise.reject("SET_USER_ERROR", e.getMessage(), e);
    }
  }

  @ReactMethod
  public void startRecording(Promise promise) {
    try {
      // Note: Requires Activity context for permissions
      promise.resolve(false); // Would need Activity context
    } catch (Exception e) {
      promise.reject("RECORDING_ERROR", e.getMessage(), e);
    }
  }

  @ReactMethod
  public void stopRecording(Promise promise) {
    try {
      UXCamSDK.getInstance().stopVideoRecording();
      promise.resolve(true);
    } catch (Exception e) {
      promise.reject("RECORDING_ERROR", e.getMessage(), e);
    }
  }

  @ReactMethod
  public void endSession(Promise promise) {
    try {
      UXCamSDK.getInstance().endSession();
      promise.resolve(true);
    } catch (Exception e) {
      promise.reject("SESSION_ERROR", e.getMessage(), e);
    }
  }

  @ReactMethod
  public void restartSession(Promise promise) {
    try {
      UXCamSDK.getInstance().restartSession();
      promise.resolve(true);
    } catch (Exception e) {
      promise.reject("SESSION_ERROR", e.getMessage(), e);
    }
  }
}
